src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs

```diff
@@ -0,0 +1,254 @@
+                string v2TokenIssuer = openIdConnectConfiguration.Issuer?.Replace(AadIssuerValidator.TenantIdTemplate, tenantIdFromToken);
+#endif
+
+                // comparing effectiveSigningKeyIssuer with v2TokenIssuer is required as well because of the following scenario:
+                // 1. service trusts /common/v2.0 endpoint 
+                // 2. service receives a v1 token that has issuer like sts.windows.net
+                // 3. signing key issuers will never match sts.windows.net as v1 endpoint doesn't have issuers attached to keys
+                // v2TokenIssuer is the representation of Token.Issuer (if it was a v2 issuer)
+                if (effectiveSigningKeyIssuer != tokenIssuer && effectiveSigningKeyIssuer != v2TokenIssuer)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidIssuerException(LogHelper.FormatInvariant(LogMessages.IDX40005, LogHelper.MarkAsNonPII(tokenIssuer), LogHelper.MarkAsNonPII(effectiveSigningKeyIssuer))));
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Validates the cloud instance of the signing key.
+        /// </summary>
+        /// <param name="securityKey">The <see cref="SecurityKey"/> that signed the <see cref="SecurityToken"/>.</param>
+        /// <param name="configuration">The <see cref="BaseConfiguration"/> provided.</param>
+        internal static void ValidateSigningKeyCloudInstance(SecurityKey securityKey, BaseConfiguration configuration)
+        {
+            if (securityKey == null)
+                return;
+
+            if (configuration is not OpenIdConnectConfiguration openIdConnectConfiguration)
+                return;
+
+            JsonWebKey matchedKeyFromConfig = GetJsonWebKeyBySecurityKey(openIdConnectConfiguration, securityKey);
+            if (matchedKeyFromConfig != null && matchedKeyFromConfig.AdditionalData.TryGetValue(AadIssuerValidatorConstants.CloudInstanceNameKey, out object value))
+            {
+                string signingKeyCloudInstanceName = value as string;
+                if (string.IsNullOrWhiteSpace(signingKeyCloudInstanceName))
+                    return;
+
+                if (openIdConnectConfiguration.AdditionalData.TryGetValue(AadIssuerValidatorConstants.CloudInstanceNameKey, out object configurationCloudInstanceNameObjectValue))
+                {
+                    string configurationCloudInstanceName = configurationCloudInstanceNameObjectValue as string;
+                    if (string.IsNullOrWhiteSpace(configurationCloudInstanceName))
+                        return;
+
+                    if (!string.Equals(signingKeyCloudInstanceName, configurationCloudInstanceName, StringComparison.Ordinal))
+                        throw LogHelper.LogExceptionMessage(
+                            new SecurityTokenInvalidCloudInstanceException(
```

Comment:
this is the only line that changes in this method vs in AadValidationParametersExtension